javascript/pkce_oauth/server/index.js

```diff
@@ -0,0 +1,174 @@
+const express = require('express');
+const cookieParser = require('cookie-parser');
+const sha256 = require('js-sha256');
+const request = require('request');
+const functions = require('firebase-functions');
+
+const challengeMethod = "S256";
+const authorizationGrantType = "authorization_code";
+const refreshGrantType = "refresh_token";
+
+const clientId = process.env.client_id;
+const clientSecret = process.env.client_secret;
+const redirectUri = process.env.redirect_uri;
+const frontEndLocation = process.env.front_end_location;
+const tokenExchangeEndpoint = process.env.token_exchange_endpoint;
+
+const stateCache = {};
+const tokenCache = {};
+
+const app = express();
+app.use(cookieParser());
+
+// Base endpoint. This should return our html.
+app.get('/', (req, res) => {
+    console.log('In /');
+
+    request(frontEndLocation).pipe(res);
+});
+
+// First step.
+// We need to generate state and code verifiers for this client.
+// We then return the info the client needs to hit the auth server.
+app.get('/first-step-auth', (req, res) => {
+    console.log('In first-step-auth');
+
+    let result = {};
```

Comment:
[Minor] It doesn't look like this needs to be defined as a map that you later update. It looks like you could just do
```js
const result = {
    code_challenge: codeChallenge,
    client_id: clientId,
    ...
}
```